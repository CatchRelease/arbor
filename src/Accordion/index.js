import styled from '@emotion/styled';

import Box from '../Box';

const Accordion = styled(Box)`
  border: 1px solid ${props => props.theme.colors.border.default};
  border-radius: ${props => props.theme.borderRadius.large};
`;

export default Accordion;
