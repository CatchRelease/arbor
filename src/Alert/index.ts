export { default as BannerAlert } from './BannerAlert';
export { default as CardAlert } from './CardAlert';
export { default as InlineAlert } from './InlineAlert';
export { default as PaneAlert } from './PaneAlert';
export { default as ToastAlert } from './ToastAlert';
