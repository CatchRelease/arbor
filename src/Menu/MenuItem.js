/** @jsx jsx */
import { jsx } from '@emotion/core';
import React from 'react';
import PropTypes from 'prop-types';
import { hideVisually } from 'polished';

import { ENTER_KEY, SPACEBAR } from '../constants';
import StyledMenuItem from './StyledMenuItem';
import Text from '../Text';
import Icon from '../Icon';
import Flex from '../Flex';

const MIN_WIDTH = '120px';
const PADDING = '8px';

class MenuItem extends React.PureComponent {
  componentDidMount() {
    document.addEventListener('keydown', this.onKeyDown, true);
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', this.onKeyDown, true);
  }

  onKeyDown = e => {
    const { key } = e;
    const { focused } = this.props;

    if (!focused) {
      return;
    }

    if (key === ENTER_KEY || key === SPACEBAR) {
      e.preventDefault();
      this.onSelect();
    }
  };

  onSelect = () => {
    const { onSelect, value } = this.props;
    onSelect(value);
  };

  render() {
    const {
      selected,
      baseColor,
      iconName,
      id,
      label,
      secondaryLabel,
      name,
      value,
      focused
    } = this.props;

    return (
      <StyledMenuItem
        {...{
          baseColor,
          selected,
          focused,
          role: 'option',
          'aria-selected': selected,
          alignItems: 'center',
          height: '40px',
          onClick: this.onSelect,
          onKeyPress: this.onKeyDown
        }}
      >
        <input {...{ id, name, value, css: hideVisually(), type: 'hidden' }} />
        <Flex width="100%" justifyContent="space-between">
          <Flex
            flex="1 0 auto"
            maxWidth={
              secondaryLabel ? `calc(100% - ${MIN_WIDTH} - ${PADDING})` : '100%'
            }
          >
            {iconName && (
              <Icon
                name={iconName}
                mr="small"
                color={baseColor}
                fontSize="16px"
              />
            )}
            <Text
              as="span"
              color="inherit"
              minWidth={MIN_WIDTH}
              fontSize="size4"
              overflow="hidden"
            >
              {label}
            </Text>
          </Flex>
          {secondaryLabel && (
            <Flex pl={PADDING} overflow="hidden" flex="0 1 auto">
              <Text
                as="span"
                fontSize="size3"
                css={theme => ({
                  // Put primary and secondary labels on the same baseline
                  lineHeight: theme.fontSizes.size4
                })}
                color="text.muted"
                maxWidth="100%"
              >
                {secondaryLabel}
              </Text>
            </Flex>
          )}
        </Flex>
      </StyledMenuItem>
    );
  }
}

MenuItem.propTypes = {
  /**
   * Indicates whether or not the menu item is selected
   */
  selected: PropTypes.bool,

  /**
   * Callback to run when a selection is made. This is generally intended for
   * internal use with the Menu component.
   */
  onSelect: PropTypes.func,

  /**
   * Main label text for the menu item
   */
  label: PropTypes.string.isRequired,

  /**
   * Secondary label text for the menu item
   */
  secondaryLabel: PropTypes.string,

  /**
   * HTML input name property for the input field
   */
  name: PropTypes.string.isRequired,

  /**
   * HTML input value property for the input field
   */
  value: PropTypes.string.isRequired,

  /**
   * Base color to use for the menu item
   */
  baseColor: PropTypes.oneOf(['blue', 'grey', 'green', 'red', 'bronze']),

  /**
   * Name of icon to use for the menu item
   */
  iconName: PropTypes.string,

  /**
   * HTML ID for the menu item
   */
  id: PropTypes.string.isRequired,

  /**
   * Whether or not the current menu item is focused
   */
  focused: PropTypes.bool
};

MenuItem.defaultProps = {
  baseColor: 'text.default',
  iconName: null,
  secondaryLabel: null,
  selected: false,
  onSelect: () => null,
  focused: false
};

export default MenuItem;
