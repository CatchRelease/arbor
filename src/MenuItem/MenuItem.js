import PropTypes from 'prop-types';

import StyledMenuItem from './StyledMenuItem';

const MenuItem = ({ innerRef, children, ...props }) => (
  <StyledMenuItem ref={innerRef} {...props}>
    {children}
  </StyledMenuItem>
);

MenuItem.propTypes = {
  /**
   * Content to render inside the menu item
   * */
  children: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.node)
  ]).isRequired,

  innerRef: PropTypes.func
};

MenuItem.defaultProps = {
  innerRef: undefined
};

export default MenuItem;
