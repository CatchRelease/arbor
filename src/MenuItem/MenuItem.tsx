import { FC, Ref } from 'react';

import StyledMenuItem from './StyledMenuItem';

type Props = {
  innerRef?: Ref<unknown>;
};

const MenuItem: FC<Props> = ({ innerRef, children, ...props }) => (
  <StyledMenuItem ref={innerRef} {...props}>
    {children}
  </StyledMenuItem>
);

export default MenuItem;
