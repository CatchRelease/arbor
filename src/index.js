export * from './theme';
export { default as Box } from './Box';
export { default as Button } from './Button';
export { default as Card } from './Card';
export { default as CardActions } from './CardActions';
export { default as CardPreview } from './CardPreview';
export { default as CardRow } from './CardRow';
export { default as Checkbox } from './Checkbox';
export { default as Dropdown } from './Dropdown';
export { default as Flex } from './Flex';
export { default as Grid } from './Grid';
export { default as Heading } from './Heading';
export { default as Icon } from './Icon';
export { default as Input } from './Input';
export { default as Label } from './Label';
export { default as Link } from './Link';
export { default as Menu } from './Menu';
export { default as Pagination } from './Pagination';
export { default as Paragraph } from './Paragraph';
export { default as Popover } from './Popover';
export { default as Text } from './Text';
export { default as Tooltip } from './Tooltip';

export { default as reset } from './reset';
export { default as theme } from './theme';
