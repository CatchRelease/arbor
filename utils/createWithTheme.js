import React from 'react';
import { CacheProvider } from '@emotion/core';
import createCache from '@emotion/cache';
import { create } from 'react-test-renderer';
import { ThemeProvider } from 'emotion-theming';
import theme from '../src/theme';

const myCache = createCache();
myCache.compat = true;

export default (Component) =>
  create(
    <CacheProvider value={myCache}>
      <ThemeProvider theme={theme}>{Component}</ThemeProvider>
    </CacheProvider>
  );
