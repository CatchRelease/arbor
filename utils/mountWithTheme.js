import { mount } from 'enzyme';
import { ThemeProvider } from '@emotion/react';
import theme from '../src/theme';

export default (Component, mountOptions = {}) => {
  const wrappedComponent = (
    <ThemeProvider theme={theme}>{Component}</ThemeProvider>
  );

  const mounted = mount(wrappedComponent, mountOptions);

  return mounted.find(Component.type);
};
