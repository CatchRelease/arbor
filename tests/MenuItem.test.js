import React from 'react';
import mountWithTheme from './utils/mountWithTheme';
import createWithTheme from './utils/createWithTheme';

import { ENTER_KEY, SPACEBAR } from '../src/constants';
import MenuItem from '../src/Menu/MenuItem';

describe('<MenuItem />', () => {
  const noop = () => null;
  const baseProps = {
    label: 'Ready to Clear',
    baseColor: 'grey',
    name: 'clearance_state',
    value: 'ready_to_clear',
    id: 'menu-item_clearance_state_ready_to_clear'
  };

  function createWithProps(additionalProps) {
    const props = Object.assign({}, baseProps, additionalProps);
    return createWithTheme(<MenuItem {...{ ...props }} />);
  }

  function mountWithProps(additionalProps) {
    const props = Object.assign({}, baseProps, additionalProps);
    return mountWithTheme(<MenuItem {...{ ...props }} />);
  }

  it('renders a MenuItem properly', () => {
    const menuItem = createWithProps();

    expect(menuItem).toMatchSnapshot();
  });

  it('renders a selected MenuItem properly', () => {
    const menuItem = createWithProps({ selected: true });

    expect(menuItem).toMatchSnapshot();
  });

  describe('onClick', () => {
    it('calls the onSelect callback with the value', () => {
      const onSelect = jest.fn();
      const value = 'ready_to_clear';
      const menuItem = mountWithProps({ onSelect, value });

      menuItem.simulate('click');

      expect(onSelect).toHaveBeenCalledWith(value);
    });
  });

  describe('keydown', () => {
    const value = 'ready_to_clear';

    const behavesLikeKeyboardSelector = key => {
      context('menu item is focused', () => {
        it('calls the onSelect callback', () => {
          const onSelect = jest.fn();
          const menu = mountWithProps({ onSelect, value, focused: true });

          menu.instance().onKeyDown({ key, preventDefault: noop });

          expect(onSelect).toHaveBeenCalledWith(value);
        });

        it('prevents the default keydown event', () => {
          const menu = mountWithProps({ focused: true });
          const preventDefault = jest.fn();

          menu.instance().onKeyDown({ key, preventDefault });

          expect(preventDefault).toHaveBeenCalled();
        });
      });

      context('menu item is not focused', () => {
        it('does not call the onSelect callback', () => {
          const onSelect = jest.fn();
          const menu = mountWithProps({ onSelect, value, focused: false });

          menu.instance().onKeyDown({ key });

          expect(onSelect).not.toHaveBeenCalled();
        });
      });
    };

    context('Enter key', () => {
      behavesLikeKeyboardSelector(ENTER_KEY);
    });

    context('Spacebar', () => {
      behavesLikeKeyboardSelector(SPACEBAR);
    });

    context('Other key than Enter or space', () => {
      context('menu item is focused', () => {
        it('does not call the onSelect callback', () => {
          const onSelect = jest.fn();
          const menu = mountWithProps({ onSelect, value, focused: true });

          menu.instance().onKeyDown({ key: 'ArrowDown' });

          expect(onSelect).not.toHaveBeenCalled();
        });
      });
    });
  });
});
