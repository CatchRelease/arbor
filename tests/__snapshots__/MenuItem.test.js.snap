// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuItem /> renders a MenuItem properly 1`] = `
.emotion-4 {
  box-sizing: border-box;
  height: 40px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #25272D;
  padding: 0 1rem;
}

.emotion-4:hover {
  background: #F8F9F9;
}

.emotion-4:active {
  background: #F2F2F2;
}

.emotion-4:focus {
  background: #F8F9F9;
}

.emotion-0 {
  color: grey;
  font-size: 16px;
  margin-right: 0.75rem;
}

.emotion-2 {
  font-family: Motiva Sans;
  font-size: 0.875rem;
}

<div
  aria-selected={false}
  className="emotion-4 emotion-5"
  height="40px"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="Menuitem"
  selected={false}
>
  <input
    css={
      Object {
        "border": "0",
        "clip": "rect(0 0 0 0)",
        "clipPath": "inset(50%)",
        "height": "1px",
        "margin": "-1px",
        "overflow": "hidden",
        "padding": "0",
        "position": "absolute",
        "whiteSpace": "nowrap",
        "width": "1px",
      }
    }
    id="menu-item_clearance_state_ready_to_clear"
    name="clearance_state"
    type="hidden"
    value="ready_to_clear"
  />
  <i
    className="ar ar-status emotion-0 emotion-1"
    color="grey"
    fontSize="16px"
  />
  <span
    className="emotion-2 emotion-3"
    fontSize="size2"
  >
    Ready to Clear
  </span>
</div>
`;

exports[`<MenuItem /> renders a selected MenuItem properly 1`] = `
.emotion-0 {
  color: grey;
  font-size: 16px;
  margin-right: 0.75rem;
}

.emotion-2 {
  font-family: Motiva Sans;
  font-size: 0.875rem;
}

.emotion-4 {
  box-sizing: border-box;
  height: 40px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #25272D;
  padding: 0 1rem;
  color: #0F0F10;
  font-weight: 700;
  padding-right: 0.5rem;
}

.emotion-4:hover {
  background: #F8F9F9;
}

.emotion-4:active {
  background: #F2F2F2;
}

.emotion-4:focus {
  background: #F8F9F9;
}

<div
  aria-selected={true}
  className="emotion-4 emotion-5"
  height="40px"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="Menuitem"
  selected={true}
>
  <input
    css={
      Object {
        "border": "0",
        "clip": "rect(0 0 0 0)",
        "clipPath": "inset(50%)",
        "height": "1px",
        "margin": "-1px",
        "overflow": "hidden",
        "padding": "0",
        "position": "absolute",
        "whiteSpace": "nowrap",
        "width": "1px",
      }
    }
    id="menu-item_clearance_state_ready_to_clear"
    name="clearance_state"
    type="hidden"
    value="ready_to_clear"
  />
  <i
    className="ar ar-status emotion-0 emotion-1"
    color="grey"
    fontSize="16px"
  />
  <span
    className="emotion-2 emotion-3"
    fontSize="size2"
  >
    Ready to Clear
  </span>
</div>
`;
